### ISSUE TITLE
So many false positives that it's a challenge to perform reviews

### ISSUE THREAD

**Issue by AaronBallman** [google/llvm-premerge-checks#91]
I love the idea of running clang-format and clang-tidy checks in an automated way as part of our phab integration, but the current state of this makes it incredibly frustrating to provide reviews because there are far too many false positives reported.

For instance, in this (https://reviews.llvm.org/D72221) review we get clang-tidy diagnostics that mirror a clang diagnostic about an unknown attribute `patchable_function_entry` -- that attribute is unknown to the clang instance running because it is being added in the patch under review, so all 11 instances of that diagnostic are incorrect.

Also (on that same review), it is complaining about clang-format for comments in the test files but we have never required users to follow the 80-col limits for comments in test code (because the expects-whatever lines can get very onerous to try to wrap properly). Others have pointed out that it frequently gets formatting wrong for `BuildMI` calls, so this may not be limited to just test code.

These sort of non-issues have happened on almost every review I've participated in since this bot came online and I would appreciate if the bot could be disabled until these sort of issues are resolved.

**Comment by MaskRay** [comment]
It also seems that `merge_guards_bot` does not understand patch dependencies. A dependent patch can likely cause failures.

**Comment by joker-eph** [comment]
Seems like a duplicate from #90 and #92 ?

**Comment by ChristianKuehnel** [comment]
@metaflow can you please take a look?

**Comment by metaflow** [comment]
I am going to disable the tidy check for now until we find a solution.

**Comment by metaflow** [comment]
cac68d0d60da066feabfef65d326729e2d9bc424

**Comment by AaronBallman** [comment]
> I am going to disable the tidy check for now until we find a solution.

Thank you for this!

**Comment by metaflow** [comment]
Looking at https://reviews.llvm.org/D72103: clang-tidy-diff produces reasonable command for clang-tidy
```
'clang-tidy',
'-line-filter=\'[{"name":"clang/lib/Sema/SemaTemplate.cpp","lines":[[4303,4319]]}, "name":"clang/test/SemaTemplate/instantiate-var-template.cpp","lines":[[40,103]]}]\'',
'clang/lib/Sema/SemaTemplate.cpp',
'clang/test/SemaTemplate/instantiate-var-template.cpp'
```
but clang-tidy reports lines outside of this ranges.

**Comment by metaflow** [comment]
Apparently it's a lit-test and it should not be fed to clang-tidy.

**Comment by metaflow** [comment]
My current plan is:
- exclude all files that are not present in `compile_command.json`
- print warnings inline only for some subset of projects that opt-in (there will be a list somewhere in this repo), for everyone else clang tidy will still show failure and link to the report file but it will not be so intrusive
Thoughts?

**Comment by joker-eph** [comment]
Will you be able to restrict to the hunks that are touched in the patch?

**Comment by njames93** [comment]
What about if a warning generates a fix-it in code that isn't touched in the patch?

**Comment by njames93** [comment]
Still getting artefacts. in [D73098](https://reviews.llvm.org/D73098) the patch only touches files in clang-tools-extra, yet the clang-tidy report for the patch suggested a change in a lldb header file.

**Comment by metaflow** [comment]
> Still getting artefacts. in [D73098](https://reviews.llvm.org/D73098) the patch only touches files in clang-tools-extra, yet the clang-tidy report for the patch suggested a change in a lldb header file.

Might it be that master branch was broken at that time? I have restarted build for your change and it looks fine now https://reviews.llvm.org/D73098#1835547

**Comment by njames93** [comment]
I saw that, but if master was broken, how come the actual build and unit tests all ran normally

**Comment by metaflow** [comment]
Closing now, please reopen or create another issue.

**Comment by metaflow** [comment]
Hi @AaronBallman , I have added "**/test" to be ignored by clang-format.
Sorry for missing some of the initial issues as discussion diverged. I really encourage you to persist on issues that bothers you and maybe reopen them if they are not fixed properly.
If you think it's in a good state now, please close the issue.

**Comment by AaronBallman** [comment]
> Hi @AaronBallman , I have added "**/test" to be ignored by clang-format.
> Sorry for missing some of the initial issues as discussion diverged. I really encourage you to persist on issues that bothers you and maybe reopen them if they are not fixed properly.
> If you think it's in a good state now, please close the issue.

Thank you for the additional work on this! I'm going to close on the assumption we're in a good state (nothing has jumped out at me in the past day or so) and I'll reopen or open a new issue if I spot something else.
